mm2src/gui_storage/src/account/storage/sqlite_storage.rs

```diff
@@ -0,0 +1,580 @@
+        &self,
+    ) -> AccountStorageResult<BTreeMap<AccountId, AccountWithEnabledFlag>> {
+        let conn = self.lock_conn_mutex()?;
+        let enabled_account_id = AccountId::from(Self::load_enabled_account_id_or_err(&conn)?);
+
+        let mut found_enabled = false;
+        let accounts = Self::load_accounts(&conn)?
+            .into_iter()
+            .map(|(account_id, account_info)| {
+                let enabled = account_id == enabled_account_id;
+                found_enabled |= enabled;
+                Ok((account_id, AccountWithEnabledFlag { account_info, enabled }))
+            })
+            .collect::<AccountStorageResult<BTreeMap<_, _>>>()?;
+
+        // If `AccountStorage::load_enabled_account_id_or_err` returns an `AccountId`,
+        // then corresponding account must be in `AccountTable`.
+        if !found_enabled {
+            return MmError::err(AccountStorageError::unknown_account_in_enabled_table(
+                enabled_account_id,
+            ));
+        }
+        Ok(accounts)
+    }
+
+    async fn load_enabled_account_id(&self) -> AccountStorageResult<EnabledAccountId> {
+        let conn = self.lock_conn_mutex()?;
+        Self::load_enabled_account_id_or_err(&conn)
+    }
+
+    async fn load_enabled_account_with_coins(&self) -> AccountStorageResult<AccountWithCoins> {
+        let conn = self.lock_conn_mutex()?;
+        let account_id = AccountId::from(Self::load_enabled_account_id_or_err(&conn)?);
+
+        Self::load_account_with_coins(&conn, &account_id)?
+            .or_mm_err(|| AccountStorageError::unknown_account_in_enabled_table(account_id))
+    }
+
+    async fn enable_account(&self, enabled_account_id: EnabledAccountId) -> AccountStorageResult<()> {
+        let conn = self.lock_conn_mutex()?;
+
+        // First, check if the account exists.
+        let account_id = AccountId::from(enabled_account_id);
+        if !Self::account_exists(&conn, &account_id)? {
+            return MmError::err(AccountStorageError::NoSuchAccount(account_id));
+        }
+
+        // Remove the previous enabled account by clearing the table.
+        SqlDelete::new(&conn, enabled_account_table::TABLE_NAME)?.delete()?;
+
+        let mut sql_insert = SqlInsert::new(&conn, enabled_account_table::TABLE_NAME);
+
+        let (account_type, account_idx) = enabled_account_id.to_sql_pair();
+        sql_insert
+            .column(enabled_account_table::ACCOUNT_TYPE, account_type)?
+            .column(enabled_account_table::ACCOUNT_IDX, account_idx)?;
+
+        sql_insert.insert()?;
+        Ok(())
+    }
+
+    async fn upload_account(&self, account: AccountInfo) -> AccountStorageResult<()> {
+        let conn = self.lock_conn_mutex()?;
+
+        // First, check if the account doesn't exist.
+        if Self::account_exists(&conn, &account.account_id)? {
+            return MmError::err(AccountStorageError::AccountExistsAlready(account.account_id));
+        }
```

Comment:
Is this needed? doesn't insert return an error if an entry already exists? it seems that this is an extra call to the DB that is not needed if that's the case.

Reply:
Then we need to determine why the SQLite command fails. I even don't know how we can do it.
> doesn't insert return an error if an entry already exists?

Reply:
> Then we need to determine why the SQLite command fails. I even don't know how we can do it.

Actually you can do that, [see this](https://docs.rs/libsqlite3-sys/0.25.0/libsqlite3_sys/struct.Error.html)

Reply:
Oh you're totally right.
> Actually you can do that, [see this](https://docs.rs/libsqlite3-sys/0.25.0/libsqlite3_sys/struct.Error.html)

@shamardy @ozkanonur I forgot the reason why I decided not to declare primary and foreign keys at the beginning.
[SQLite documentation](https://www.sqlite.org/lang_createtable.html):
> Each row in a table with a primary key must have a unique combination of values in its primary key columns. For the purposes of determining the uniqueness of primary key values, NULL values are considered distinct from all other values, including other NULLs.

Since I designed the tables the way that `account_idx` and `device_pubkey` can be NULL depending on the `account_type` value, the Primary key constraint will not work.
For example, for the following table
```sql
CREATE TABLE foo (
  account_type NUMBER NOT NULL,
  account_idx NUMBER,
  device_pubkey VARCHAR(20),
  CONSTRAINT account_id PRIMARY KEY (account_type, account_idx, device_pubkey)
);
```
it's possible to insert multiple `account_type=1, account_idx=NULL, device_pubkey=NULL` elements, since NULL is considered distinct from other NULLs.

But... the same applies to `UNIQUE` constraint. So, I made a bug.
As a solution I consider using the following scheme for `account_type`, `account_idx`, `device_pubkey`:
1) If `account_type == 0` (Iguana), then `account_idx=0` (default), `device_pubkey=""` (default), where `account_idx` and `device_pubkey` will be just ignored
2) If `account_type == 1` (HD), then `account_idx=IDX`, `device_pubkey=""` (default), where `device_pubkey` will be just ignored
3) If `account_type == 1` (HW), then `account_idx=0` (default), `device_pubkey="PUBKEY"` where `account_idx` will be just ignored

In this scheme I'll be able to declare PRIMARY and FOREIGN constraints with CASCADE policy, and will parse the SQLite error.
What do you think?

Reply:
I've finished with this refactoring.

Thank you @shamardy @ozkanonur
You helped me to make this code better, more optimal, at the same time I learned a lot of new things.